Lib/pathlib/_local.py

```diff
@@ -634,6 +672,12 @@ def _filter_trailing_slash(self, paths):
                 path_str = path_str[:-1]
             yield path_str
 
+    def _join_dir_entry(self, dir_entry):
+        path_str = dir_entry.name if str(self) == '.' else dir_entry.path
```

Comment:
Could we shortcut here rather than going via `str(self)`? The change to `iterdir` means that this is recalculated for each path in the directory being iterated over, rather than only once.